// @shopify/polaris/esnext/utilities/with-app-provider.js
import React from 'react';
import hoistStatics from 'hoist-non-react-statics';
import { useI18n } from './i18n';
import { useLink } from './link';
import { useScrollLockManager } from './scroll-lock-manager';
import { useTheme } from './theme';
import { useStickyManager } from './sticky-manager';
import { useAppBridge } from './app-bridge';
export function withAppProvider() {
    return function addProvider(WrappedComponent) {
        const WithAppProvider = (props) => {
            const polaris = {
                link: useLink(),
                theme: useTheme(),
                intl: useI18n(),
                scrollLockManager: useScrollLockManager(),
                stickyManager: useStickyManager(),
                appBridge: useAppBridge(),
            };
            return <WrappedComponent {...props} polaris={polaris}/>;
        };
        WithAppProvider.displayName = `WithAppProvider(${getDisplayName(WrappedComponent)})`;
        const FinalComponent = hoistStatics(WithAppProvider, WrappedComponent);
        return FinalComponent;
    };
}
function getDisplayName(WrappedComponent) {
    return WrappedComponent.displayName || WrappedComponent.name || 'Component';
}
