// @shopify/polaris/esnext/components/Tabs/components/Item/Item.js
import React from 'react';
import styles from '../../Tabs.scss';
import { UnstyledLink } from '../../../UnstyledLink';
export class Item extends React.PureComponent {
    constructor() {
        super(...arguments);
        this.focusedNode = null;
        this.setFocusedNode = (node) => {
            this.focusedNode = node;
        };
    }
    componentDidMount() {
        const { focusedNode } = this;
        const { focused } = this.props;
        if (focusedNode && focusedNode instanceof HTMLElement && focused) {
            focusedNode.focus();
        }
    }
    componentDidUpdate() {
        const { focusedNode } = this;
        const { focused } = this.props;
        if (focusedNode && focusedNode instanceof HTMLElement && focused) {
            focusedNode.focus();
        }
    }
    render() {
        const { id, panelID, children, url, accessibilityLabel, onClick = noop, } = this.props;
        const sharedProps = {
            id,
            ref: this.setFocusedNode,
            onClick,
            className: styles.Item,
            'aria-controls': panelID,
            'aria-selected': false,
            'aria-label': accessibilityLabel,
        };
        const markup = url ? (<UnstyledLink {...sharedProps} url={url}>
        {children}
      </UnstyledLink>) : (<button {...sharedProps} type="button">
        {children}
      </button>);
        return <li>{markup}</li>;
    }
}
function noop() { }
